src/VisualStudio/Core/Test.Next/Remote/RemoteHostClientServiceFactoryTests.cs

```diff
@@ -71,6 +71,99 @@ public async Task UpdaterService()
             checksumUpdater.Shutdown();
         }
 
+        private static readonly ImmutableArray<string> s_kinds = ImmutableArray.Create(
+            NavigateToItemKind.Class,
+            NavigateToItemKind.Constant,
+            NavigateToItemKind.Delegate,
+            NavigateToItemKind.Enum,
+            NavigateToItemKind.EnumItem,
+            NavigateToItemKind.Event,
+            NavigateToItemKind.Field,
+            NavigateToItemKind.Interface,
+            NavigateToItemKind.Method,
+            NavigateToItemKind.Module,
+            NavigateToItemKind.Property,
+            NavigateToItemKind.Structure);
+
+        [Fact]
+        public async Task TestStreamingServices()
+        {
+            using var workspace = CreateWorkspace();
+
+            var exportProvider = workspace.Services.SolutionServices.ExportProvider;
+            var listenerProvider = exportProvider.GetExportedValue<AsynchronousOperationListenerProvider>();
+            var globalOptions = exportProvider.GetExportedValue<IGlobalOptionService>();
+
+            var service = workspace.Services.GetRequiredService<IRemoteHostClientProvider>();
+            using var outerClient = await service.TryGetRemoteHostClientAsync(CancellationToken.None);
+
+            // add solution, change document
+            workspace.AddSolution(SolutionInfo.Create(SolutionId.CreateNewId(), VersionStamp.Default));
+            var project = workspace.AddProject("proj", LanguageNames.CSharp);
+            var document = workspace.AddDocument(project.Id, "doc.cs", SourceText.From(""));
+
+            var assetStorage = workspace.Services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
+
+            var tasks = new List<Task>();
+
+            // Kick off a ton of work in parallel to try to shake out race conditions with pinning/syncing.
+
+            // 100 outer loops that ensure we generate the same files with the same checksums at least 10 times.
+            for (var i = 0; i < 100; i++)
+            {
+                // 100 inner loops producing variants of the file.
+                for (var j = 0; j < 100; j++)
+                {
+                    tasks.Add(Task.Run(async () => await PerformSearchesAsync(service, assetStorage, document, name: "Goo" + i)));
+                }
+            }
+
+            await Task.WhenAll(tasks);
+        }
```

Comment:
sorry, one last suggestion - might be good to also assert that once everything is complete nothing is pinned

Reply:
will do :)